src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpChangeNamespaceService.cs

```diff
@@ -135,89 +192,163 @@ protected override SyntaxList<MemberDeclarationSyntax> GetMemberDeclarationsInCo
+                        CreateNameSyntax(targetNamespaceParts, aliasQualifier: null, targetNamespaceParts.Length - 1)
+                        .WithTriviaFrom(namespaceDecl.Name).WithAdditionalAnnotations(WarningAnnotation))
+                        .WithoutAnnotations(ContainerAnnotation));
             }
 
-            // We should have a single member which is a namespace declaration in this compilation unit.
-            var namespaceDeclaration = compilationUnit.DescendantNodes().OfType<NamespaceDeclarationSyntax>().Single();
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        /// <summary>
+        /// For the node specified by <paramref name="span"/> to be applicable container, it must be a namespace 
+        /// declaration or a compilation unit, contain no partial declarations and meet the following additional
+        /// requirements:
+        /// 
+        /// - If a namespace declaration:
+        ///    1. It doesn't contain or is nested in other namespace declarations
+        ///    2. The name of the namespace is valid (i.e. no errors)
+        ///
+        /// - If a compilation unit (i.e. <paramref name="span"/> is empty), there must be no namespace declaration
+        ///   inside (i.e. all members are declared in global namespace)
+        /// </summary>
+        protected override async Task<SyntaxNode> TryGetApplicableContainerFromSpanAsync(Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var compilationUnit = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false) as CompilationUnitSyntax;
+            SyntaxNode container = null;
 
-            // Move everything to global namespace
-            if (IsGlobalNamespace(targetNamespaceParts))
+            // Empty span means that user wants to move all types declared in the document to a new namespace.
+            // This action is only supported when everything in the document is declared in global namespace,
+            // which we use the number of namespace declaration nodes to decide.
+            if (span.IsEmpty)
             {
-                var (namespaceOpeningTrivia, namespaceClosingTrivia) = 
-                    GetOpeningAndClosingTriviaOfNamespaceDeclaration(namespaceDeclaration);
-                var members = namespaceDeclaration.Members;
-                var eofToken = compilationUnit.EndOfFileToken
-                    .WithAdditionalAnnotations(WarningAnnotation);
-
-                // Try to preserve trivia from original namesapce declaration.
-                // If there's any member inside the declaration, we attach them to the 
-                // first and last member, otherwise, simply attach all to the EOF token.
-                if (members.Count > 0)
+                var hasNamespaceDecl = compilationUnit.DescendantNodes(IsCompilationUnitOrNamespaceDeclaration)
+                    .OfType<NamespaceDeclarationSyntax>().Any();
```

Comment:
you're asking if any descendants of the CompilationUnit are themselve a CompilationUnit.  Seems odd.

Reply:
No, I'm asking if any descendants is NamespaceDeclaration

Reply:
Fixed